### ISSUE TITLE
Include Confirm sigs of Inputs in transaction

### ISSUE THREAD

**Issue by AdityaSripal** [FourthState/plasma-mvp-rootchain#18]
To solve the issue brought up here: https://ethresear.ch/t/plasma-vulnerabiltity-sybil-txs-drained-contract/1654

Confirm sigs of the inputs must be included in transaction. Transaction structure must now be modified

`[Blknum1, TxIndex1, Oindex1, [ConfirmSig1-1, ConfirmSig1-2],
Blknum2, TxIndex2, Oindex2, [ConfirmSig2-1, ConfirmSig2-2],
NewOwner, Denom1,
NewOwner, Denom2]`

Here ConfirmSig1-1 is the first confirm sig from the first input, confirm sig is the second confirm sig from the first input, etc.
If there is no confirm sig for the given position, replace with a zero-bytes.

Changes to be made:
Deposit, StartExit and ChallengeExit must be changed to handle new transaction structure

Note Possible Confusion: We are not including confirm sig for transaction itself, we are including confirm sigs of the **INPUTS**

Consider Transactions A -> B -> C:
Only one input and output for each transaction for simplicity sake.

The Transaction B->C will include the confirm sig A sent to B for the transaction A->B.

**Comment by hamdiallam** [comment]
I dont think we should change the transaction structure. The leaf can just be keccak(txBytes, sigs, confirmsigs)

**Comment by AdityaSripal** [comment]
@colin-axner You are right. Will change

@hamdiallam I think we need to change it because the confirm sigs are part of the message. In fact, because the entire transaction is included in the merkle tree, we actually will probably have to modify this further by including the signatures in the transaction structure as well.

**Comment by hamdiallam** [comment]
@AdityaSripal  the txBytes is not included in the merkle tree. It is the hash of the bytes that is the leaf in the merkel tree

**Comment by colin-axner** [comment]
Yeah but the txBytes will include sigs and confirmsig in the message of the transaction on the child chain.

**Comment by hamdiallam** [comment]
Exactly, I think we should always make the txByes as simple as we can on the rootchain. Playing around with the way the hash is generated is better than including those inputs in the txBytes. They can be included the child chain though

**Comment by colin-axner** [comment]
@hamdiallam I'm pretty sure we would need to fork tendermint and manually keep our forked version up to date (this may need to be done anyways to update the hash function). Also adjusting tendermint to exclude hashing some parts of the transaction could lead to further issues.

We can try changing the implementation in tendermint and adjust the decoding on the root contract as a backup option

@AdityaSripal thoughts?

**Comment by hamdiallam** [comment]
Ah, i see the issue. currently, the leaves of the merkle tree are 32byte hashes. We could change this though if tendermint have the merkle tree leaves be hashes

**Comment by AdityaSripal** [comment]
@hamdiallam Doesn't seem like the sdk gives us that much flexibility. If we wanted to do that we would have to make major changes to the sdk. I would resist doing that. I think it's easier to add 4 extra fields to the txBytes.

**Comment by AdityaSripal** [comment]
We need to rlp encode the message and then attach the sigs at the end to make it easier to validate on the rootchain.

**Comment by AdityaSripal** [comment]
Also, we need to include the original denominations of the inputs. This is because we want to recover the original transaction bytes that created a deposit hash. This is only possible if we include the input denomination as part of the transaction.

**Comment by colin-axner** [comment]
@AdityaSripal I think you were originally correct about using confirmSig1-1, confirmSig1-2. While each output has separate confirmSigs, having two confirmSigs for one output is necessary when both inputs came from different addresses.

**Comment by hamdiallam** [comment]
Yeah, this can all be one sigs bytes variable that gets split up on the rootchain contract
